src/safeposix/syscalls/fs_calls.rs

```diff
@@ -910,6 +909,421 @@ impl Cage {
+                                    interface::removefile(sysfilename).unwrap();
+            
+                                } 
+                            }
+                        },
+                        Inode::Dir(ref mut dir_inode_obj) => {
+                            dir_inode_obj.refcount -= 1;
+
+                            //if there are still references to the file, then do nothing to the file
+                            if dir_inode_obj.refcount != 0 {
+                                return 0;
+                            }
+                            if fobjtable.contains_key(&inodenum) {
+                                //close and delete the file in the object table
+                                fobjtable.get(&inodenum).unwrap().close().unwrap();
+                                fobjtable.remove(&inodenum);
+                            }
+                            if dir_inode_obj.refcount == 0 {
+                                if dir_inode_obj.linkcount == 0  {
+                                    mutmetadata.inodetable.remove(&inodenum);
+                                    let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
+                                    interface::removefile(sysfilename).unwrap();
+            
+                                } 
+                            }
+                        },
+                        Inode::CharDev(ref mut char_inode_obj) => {
+                            char_inode_obj.refcount -= 1;
+
+                            //if there are still references to the file, then do nothing to the file
+                            if char_inode_obj.refcount != 0 {
+                                return 0;
+                            }
+
+                            if fobjtable.contains_key(&inodenum) {
+                                //close and delete the file in the object table
+                                fobjtable.get(&inodenum).unwrap().close().unwrap();
+                                fobjtable.remove(&inodenum);
+                            }
+                            if char_inode_obj.refcount == 0 {
+                                if char_inode_obj.linkcount == 0  {
+                                    mutmetadata.inodetable.remove(&inodenum);
+                                    let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
+                                    interface::removefile(sysfilename).unwrap();
+            
+                                } 
+                            }
+                        },
+                        Inode::Pipe(_) | Inode::Socket(_) => {panic!("How did you get by the first filter?");},
+                    }
+                },
+            }
+            // if this is a part of the close syscall, then the fd table instance of the fd is removed
+            if is_close_syscall {
+                self.filedescriptortable.write().unwrap().remove(&fd);
+            }
```

Comment:
The file descriptor should be removed whether it is a stream or socket or pipe as well, put it outside the if let

Reply:
Also whether refcount is 0 or not